File: src/playback/playbackTelemetry.js
import { sanitizeTelemetryFields } from './telemetrySanitizer.js';

const DEFAULT_MAX_ATTEMPTS = 3;
const DEFAULT_QUEUE_LIMIT = 100;
const DEFAULT_RETRY_MS = 1_000;

function defaultWait(delay) {
    return new Promise((resolve) => setTimeout(resolve, delay));
}

export function createPlaybackTelemetry({
    client,
    pipeline,
    wait = defaultWait,
    retryMs = DEFAULT_RETRY_MS,
    maxAttempts = DEFAULT_MAX_ATTEMPTS,
    queueLimit = DEFAULT_QUEUE_LIMIT
}) {
    const queue = [];
    let drainPromise = null;
    let droppedCount = 0;

    async function drain() {
        while (queue.length > 0) {
            const entry = queue[0];
            try {
                await client.postNoContent('/ClientLog/PlaybackPipeline', entry.payload);
                queue.shift();
            } catch {
                entry.attempts += 1;
                if (entry.attempts >= maxAttempts) {
                    queue.shift();
                    droppedCount += 1;
                } else {
                    await wait(retryMs);
                }
            }
        }
    }

    function startDrain() {
        if (drainPromise) return;
        drainPromise = drain().finally(() => {
            drainPromise = null;
            if (queue.length > 0) startDrain();
        });
    }

    return {
        get droppedCount() {
            return droppedCount;
        },
        log(event, fields = {}) {
            if (queue.length >= queueLimit) {
                droppedCount += 1;
                return false;
            }
            const payload = sanitizeTelemetryFields(pipeline.nextEvent(event, {
                ...fields,
                telemetryDroppedCount: droppedCount
            }));
            queue.push({ payload, attempts: 0 });
            startDrain();
            return true;
        },
        flush() {
            return drainPromise ?? Promise.resolve();
        }
    };
}

File: apps/android-tv/feature/player/src/main/kotlin/app/homeflix/tv/feature/player/PlaybackTelemetry.kt
package app.homeflix.tv.feature.player

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.json.JsonObject

private const val DEFAULT_MAX_ATTEMPTS = 3
private const val DEFAULT_QUEUE_LIMIT = 100
private const val DEFAULT_RETRY_MS = 1_000L

class PlaybackTelemetry(
    private val gateway: TelemetryGateway,
    private val pipeline: PlaybackPipelineIds,
    private val scope: CoroutineScope,
    private val retryMs: Long = DEFAULT_RETRY_MS,
    private val maxAttempts: Int = DEFAULT_MAX_ATTEMPTS,
    private val queueLimit: Int = DEFAULT_QUEUE_LIMIT,
) {
    private val queue = ArrayDeque<QueuedEvent>()
    private var drainJob: Job? = null

    var droppedCount = 0
        private set

    fun log(
        event: String,
        fields: Map<String, Any?> = emptyMap(),
    ): Boolean {
        if (queue.size >= queueLimit) {
            droppedCount += 1
            return false
        }
        val payload = pipeline.nextEvent(event, fields + ("telemetryDroppedCount" to droppedCount))
        queue.addLast(QueuedEvent(payload))
        startDrain()
        return true
    }

    suspend fun flush() {
        drainJob?.join()
    }

    private fun startDrain() {
        if (drainJob?.isActive == true) return
        drainJob =
            scope.launch {
                while (queue.isNotEmpty()) {
                    val entry = queue.first()
                    if (deliver(entry.payload)) {
                        queue.removeFirst()
                    } else {
                        entry.attempts += 1
                        if (entry.attempts >= maxAttempts) {
                            queue.removeFirst()
                            droppedCount += 1
                        } else {
                            delay(retryMs)
                        }
                    }
                }
            }
    }

    private suspend fun deliver(payload: JsonObject): Boolean =
        try {
            gateway.logPipelineEvent(payload)
            true
        } catch (failure: CancellationException) {
            throw failure
        } catch (_: Exception) {
            false
        }

    private class QueuedEvent(
        val payload: JsonObject,
    ) {
        var attempts = 0
    }
}
